src/app/CommandHandler.cpp

```diff
@@ -284,33 +341,62 @@ void CommandHandler::DecrementHoldOff()
         }
         else if (!IsGroupRequest())
         {
-            CHIP_ERROR err = SendCommandResponse();
+            CHIP_ERROR err = InitiateSendingCommandResponses();
             if (err != CHIP_NO_ERROR)
             {
                 ChipLogError(DataManagement, "Failed to send command response: %" CHIP_ERROR_FORMAT, err.Format());
             }
         }
     }
 
+    if (mState == State::AwaitingResponse)
+    {
+        // If we are AwaitingResponse from command that was successfully sent out, the responsibility to call Close()
+        // is now expected to be completed by the ExchangeDelegate callbacks of CommandHandler. In order to get those
+        // callbacks we have to set ourselves as the delegate and return.
+        mExchangeCtx->SetDelegate(this);
+        return;
+    }
     Close();
```

Comment:
This is a good candidate for

if (xx)
{
   blah;
} else {
   Close();
}

Reply:
Why is that better? I am not sure I follow. I want Close() to remain the last thing that happen in any function it is in. While there is nothing between that last else and the end of this function I find that else code block has a higher change that someone accidentally introduce something between Close() and the end of the method.

Reply:
Since the comment block was very complex/long to explain basically why the early return, I thought perhaps the above was more explicit from code reading alone. I didn't know there was an assumption "Close should remain the last thing that happens in any function". It's obviously skipped in some cases, so it becomes moot.